src/python/gudhi/representations/vector_methods.py

```diff
@@ -805,7 +815,7 @@ def fit(self, X, y=None, sample_weight=None):
             self.inertias = np.min(dist_centers, axis=0)/2
         return self
 
-    def __call__(self, measure, sample_weight=None):
+    def _transform(self, measure, sample_weight=None):
```

Comment:
The name `_transform` means that it won't appear in the doc, but we use it in the main example, that's not consistent. @martinroyer , could you remind me what the problem was with `__call__`?

Reply:
Hmm, I believe it is more similar to what the people at sklearn do, which is pertinent because these are "transformers", whereas `__call__` came from nowhere when I wrote it.
So if you ask me, unless `__call__`  is consistent with other parts of gudhi I favor the sklearn style and would just use `atol_vectoriser.transform([a])` for consistency.

Reply:
We have `__call__` in most other transformers as a shortcut for the case where we want to transform a single diagram. It is clearly not the main interface, and I don't know if it was a good idea, but I don't think it hurts that much. I would be in favor of keeping it to avoid needlessly breaking code that might be using it. We could change the example to make it use `transform` instead of `__call__` though, to advertise the main API. Would that be ok with you?
(A wider discussion about the API of transformers in gudhi is in #982, still open to comments)

Reply:
I'd like to make the case for not going that direction!

One of the things I gain from looking at code where there are `transform` and `_transform` (yes mostly in sklearn) is that I know that `transform` is going to contain most of the checks (e.g. input checks), and the real  "transform" mechanisms are found in the sub `_transform` functions. So I go through code faster because of that convention, I go directly to the level that interests me.

In this (Atol) instance it is a bit different, the `_transform` acts on a single measure and the `tranform` acts on a list of measures, which is another mechanical distinction, and probably a bad one if we do not want to hide the `_transform` on a single measure (perhaps a conversation to be had?).
But the good part is that the naming convention still shows that those two are mechanically linked. And even if I did not know the convention, I would know these functions are linked to each other because of the name, whereas `__call__` doesn't relate to `transform` at all.

So I guess my vote goes to either :
- decide that we publicly only manipulate lists of measures, therefore hide the transform on a single measure (so as is, `transform` and `_transform`),
- don't decide that but rename `_transform` to `transform_single` (or something...), not `__call__`.

(lastly `__call__` is not IDE-clickable when used someplace which is another downside!)

Reply:
There are several things discussed here, and I think considering them separately would help.
1. Fixing compatibility with recent versions of scikit-learn. This only involves passing some explicit arguments, and is the main goal of this PR. It needs to happen quickly for the next release.
2. Handling pandas output. This is essentially one new short function and an extra line of code to set a member variable, I don't mind handling it at the same time as the above, although it could make sense to add it to more classes at the same time (in a separate PR then).
3. What the interface of all those vectorization classes should be. In my opinion, whatever is decided there, it should be done in a separate PR, and handle all classes consistently at the same time, not just Atol (unless there is a good reason for the difference).

Now 3. can be subdivided into
- Should we keep providing a convenience function (not part of the standard sklearn interface) for the case where we only transform 1 diagram?
- If yes, what should its public name be? `__call__`, `transform_single` (that's going to be longer to use than `transform([x])[0]`), both...
- If as an implementation detail the core computation of `transform`/`__call__`/`transform_single` should be in a function called `_transform` (I don't care either way, I don't mind approving a PR that changes this detail (but not the API) at any time, even if it isn't very consistent between classes, if the author feels strongly about it)
- If we have a public transform_single (whatever its name), whether the examples should use it, or use the main `transform` only, to encourage using one or the other
- Assuming we decide to remove `__call__` from the documentation (either remove or rename), whether we should secretly keep it (without doc) in the code to avoid obnoxiously breaking users' code
- If IDEs that cannot handle standard magic functions like `__call__` suck (please ignore this troll)

I don't think I am really contradicting your end goal too strongly, this is mostly about the order in which to do the changes.

Reply:
Perfectly summarized!, agree on all points except that `transform_single` is clearly way superior to `transform([x])[0]` (it makes both eyes and hands bleed...)

Reply:
I renamed `_transform` as `__call__` on [e4226ea](https://github.com/GUDHI/gudhi-devel/pull/1056/commits/e4226ea14d00bdc5cc878731a4ab8d677a5445c2)